### PR TITLE
[GlobalISel][NFC] Fix LLT Propagation

### PR DESCRIPTION
**Pull request by tgymnich**
Retain LLT type information by creating new LLTs from the original LLT instead of only using the original scalar size.

This PR prepares for the [LLT FPInfo RFC](https://discourse.llvm.org/t/rfc-globalisel-adding-fp-type-information-to-llt/83349/24) where LLTs will carry additional floating point type information in addition to the scalar size.

### CODE REVIEW

**Comment by tgymnich** [comment]
> Should we long-term deprecate those vector constructors that take bits?
@tschuett That would be one of the goals of the RFC. I'll create a follow up PR containing those changes.

**Comment by aemerson** [comment]
I'm wondering if we should add a new API so that it's convenient to create a new LLT based on the existing vector LLT. Instead of having to manually call `getElementType()`. Maybe `LLT::fixed_vector(unsigned EC, LLT &OrigLLT)` that pulls the type from the original one?

**Comment by aemerson** [comment]
> I'm wondering if we should add a new API so that it's convenient to create a new LLT based on the existing vector LLT. Instead of having to manually call `getElementType()`. Maybe `LLT::fixed_vector(unsigned EC, LLT OrigLLT)` that pulls the type from the original one?

Ah, we already have that overload for the scalar variant. We can deal with this in another change.

**Review by aemerson — APPROVED** [review]

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:5352]: I think this could use LLT::divide

**Review by arsenm — APPROVED** [review]

**Comment by tgymnich** [comment]
Tracking issue: https://github.com/llvm/llvm-project/issues/119667

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp
Comment (tgymnich) [reply llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:5352]: `LLT::divide` operates on element counts not on bit sizes.

**Comment by tgymnich** [comment]
Feel free to merge this for me since I do not have commit access. In the meanwhile I already started drafting a PR that removes initialization with scalar sizes: https://github.com/llvm/llvm-project/pull/119725

**Comment by aemerson** [comment]
> Feel free to merge this for me since I do not have commit access. In the meanwhile I already started drafting a PR that removes initialization with scalar sizes: #119725

Sure, just a tip on GitHub workflow though: it's recommended you don't merge into your PR branch from main. If you need to update you should do a rebase and then force push the new branch. In most cases however rebasing isn't done either unless it's been a long time since it can also mess with the review comment UI a bit.

**Pull request merged**
